add get/set mayu config API endpoint

calvix opened the pull request:
New function to mayu/mayuctl - `config`
This allows viewing and updating mayu config via mayuctl, which means we no longer need to redeploy and build whole mayu with ansible when we want to make changes like updating vault-token.
The ideal state is that we deploy mayu and we leave it running  unless we want to update docker image.

It also helps me iterate much faster when dealing with new machines.

One thing to consider is that mayu doesn't use any kind of authorization for fetching/changing data, including this new operations.


another step is moving everything to `igniton` instead of cloud-config

ping @giantswarm/team-positive

xh3b4sd commented on client/client.go:
```diff
@@ -262,3 +264,38 @@ func (c *Client) Status(serial string) (hostmgr.Host, error) {
+	return string(body), nil
+}
+
+func (c *Client) SetConfig(conf pxemgr.Configuration) (error) {
```
`(error)` should be `error`, which probably gets fixed automatically as well when you run `gofmt`.

xh3b4sd commented on mayuctl/config.go:
```diff
@@ -0,0 +1,70 @@
+func configRun(cmd *cobra.Command, args []string) {
+
+	if configFlags.Method == "get" {
+
```
Please remove the empty lines unless they serve a real purpose.

xh3b4sd commented on mayuctl/config.go:
```diff
@@ -0,0 +1,70 @@
+)
+
+func init() {
+	configCmd.PersistentFlags().StringVar(&configFlags.Method, "method", "get","Method - 'get' or 'set', default is get. Set method requires config-file flag.")
```
Is it a conscious decision to have the actual action being a flag? I would like to align this with `gsctl`, where we have something like `gsctl create cluster`. Here for instance it would then be `mayuctl set config`. I understood we want to go with mayu in the longer term, so giving it more love would be really cool. Maybe @marians also has some opinion on this UX wise.

xh3b4sd commented on mayuctl/config.go:
```diff
@@ -0,0 +1,70 @@
+		}
+
+	} else if configFlags.Method == "set" {
+		if configFlags.ConfigFile == "" {
```
When we would have separate commands like I already suggested above, we could separate the logic a bit. The condition jungle would then not be that dark. :D

xh3b4sd commented on mayuctl/config.go:
```diff
@@ -0,0 +1,70 @@
+		} else {
+			fmt.Println(config)
+		}
+
```
Dat empty line.

xh3b4sd commented on mayuctl/config.go:
```diff
@@ -0,0 +1,70 @@
+			println("ERROR: when setting config in mayu", err)
+			os.Exit(1)
+		}
+
```
Dat empty line.

xh3b4sd commented on pxemgr/config.go:
```diff
@@ -27,14 +27,27 @@ func loadConfig(filePath string) (configuration, error) {
 
-type configuration struct {
+func saveConfig(filePath string, conf Configuration) (error) {
+
```
Dat empty line.

xh3b4sd commented on pxemgr/ipxe_handlers.go:
```diff
@@ -696,3 +698,31 @@ func (mgr *pxeManagerT) hostCoreOSVersion(r *http.Request) string {
+	err = yaml.Unmarshal(data, &conf)
+	if err != nil {
+		panic(err)
+		return
```
The return is not required if you panic. In the long run I would like to do proper logging. In our microservices we decided to do structured logging. We also agreed to prevent using `panic` whenever possible.

xh3b4sd commented on pxemgr/pxemanager.go:
```diff
@@ -212,6 +214,9 @@ func (mgr *pxeManagerT) startIPXEserver() error {
 	mgr.router.Methods("PUT").PathPrefix("/admin/host/{serial}/override").HandlerFunc(withSerialParam(mgr.override))
 
+	mgr.router.Methods("GET").PathPrefix("/admin/mayu_config").HandlerFunc(mgr.getMayuConfig)
+	mgr.router.Methods("PUT").PathPrefix("/admin/mayu_config").HandlerFunc(mgr.setMayuConfig)
```
I would like to get @marians opinion on the API scheme if that makes sense.

marians replied:
Is `PUT /admin/mayu_config` the operation used to create what can later be fetched via `GET /admin/mayu_config`? Then it's alright.

If there is anything else you want my comment on, please give me some detail. I'm lacking some context to otherwise.

xh3b4sd replied:
E.g. you had a quite strong opinion on having a slash at the end of each path, which I still not agree with, but that is another story. `mayu_config` in this case with underscore is also new to me. I did not see that before in our APIs. So it is a question to which extend we want to comply to our own rules. IMO we should apply all of your conventions to all of our projects. When we go with mayu in the long run, mayu should not be excluded.

marians replied:
I won't keep you from applying rules/guidelines from the public API to internal APIs. However I don't think we have to be as strict and as forward-thinking about the design and consistency. And also, I won't act as the rule keeper for our internal projects. I happily do this for the public API as that is customer-facing.

xh3b4sd replied:
All good. I just wanted to shout and make sure we agree on something.

calvix commented:
@xh3b4sd 
changes:
* added MaskAny (https://github.com/giantswarm/cert-operator/blob/3295c9ba52288a80703b108aa559684ff07d12a2/vendor/github.com/giantswarm/microkit/error/error.go) error function and applied to `client/client.go` ( replacing it in whole project would be future step)
* moved `defer resp.Body.Close() after error in all cases

regarding `mayuctl config -method=set` vs `mayuctl set config`, I get what you mean, but  `set` is already used for setting metadata on a specific machine and putting there also changing config seems weird.

Then you have  `mayuctl set state` which is specific to single machine and `mayuctl set config` which apply to all machines as it is global mayu config.

That is why i wanted different operation for that.

marians commented:
@calvix Did you merge this without CI passing and without a positive review? Just wondering...

calvix commented:
@marians We had discussion about this state  and i needed to merge this to be able to continue on another PR, so i asked team in lunch hangout if its ok if i merge it now and solve in another PR (because it was vendoring issue). Team say its ok so i did this exception.
